### PR TITLE
RUN-46 Reduce the launching time of the application (launch screen)

### PR DESCRIPTION
**Pull request by Qeezii**
## Description 📝

Completed tasks: 
1) **RUN-46** Reduce the launching time of the application (launch screen)

-  **RUN-101** Apply multithreading techniques to speed up application startup
Took out the logic of loading data from AppDelegate to the DataManager class.
Moved download data from server to userInteractive stream and made the download asynchronous.

-  **RUN-102** Integrate downloaded data into the app
The downloaded data is used on the library screen and on the generator screen.
Added activity indicator to the library screen and generator screen. Until the data is loaded, the activity indicator will be displayed, but after loading, the screen will update and the activity indicator will hide.

-  **RUN-103** Implement saving of uploaded data to Core Data
A folder has been created for working with the core data. It is located on the following path: Runar->Services->Memory->Core Data. There is also a CoreDataManager class with all the required methods to work with Core Data.
After the data is downloaded from the server, the available data in the core data should be checked. If the check shows that there is data in the core data, then we clear the core data by deleting all data.
Then we create the core data from the data received from the server and use it in our view controllers.

2) **RUN-98** Remove bread crumbs
The navigation path was removed, which was displayed at the top of the screen when new screens were opened in the library.

---

These are ready features, testing is required.

---

### CODE REVIEW

**Comment by GeraStupakov** [comment]
Please resolve merge conflicts

File: Runar/Runar/Application/AppDelegate.swift [hunk 1]
```diff
@@ -10,6 +10,7 @@ import UIKit
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
     var window: UIWindow?
+    let dataManager: DataManager = .init()
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Application/AppDelegate.swift:13]: @Qeezii use, please, simple init
let dataManager = DataManager()

File: Runar/Runar/Controllers/Generator flow/GenerationPopUpViewController.swift [hunk 2]
```diff
@@ -102,13 +102,17 @@ public class GenerationPopUpViewController: UIViewController {
...
-    public func setupModel(_ model: GenerationRuneModel?) {
+
+    public func setupModel(_ model: GeneratorCoreData?) {
         self.runeModel = model
-        
-        self.imageView.image = model?.image.image
-        self.header.attributedText = UILabel.getAttributedText(text: model!.title, lineHeight: 0.7)
-        self.desc.attributedText = UILabel.getAttributedText(text: model!.description, lineHeight: 1.12)
+
+        guard let imageData = model?.runeImage?.image else { return }
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Controllers/Generator flow/GenerationPopUpViewController.swift:109]: imageData or image?

File: Runar/Runar/Controllers/Generator flow/SelectRune/SelectRuneCell.swift [hunk 3]
```diff
@@ -42,12 +42,15 @@ class SelectRuneCell: UICollectionViewCell {
             runeImage.centerYAnchor.constraint(equalTo: centerYAnchor)
         ])
     }
-    
-    public func setRune(_ rune: GenerationRuneModel, _ indexPath: IndexPath) {
+
+    public func setRune(_ rune: GeneratorCoreData, _ indexPath: IndexPath) {
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Controllers/Generator flow/SelectRune/SelectRuneCell.swift:46]: @Qeezii how GeneratorCoreData matches with rune?

File: Runar/Runar/Controllers/Generator flow/SelectRune/SelectRuneCollectionView.swift [hunk 4]
```diff
@@ -56,11 +58,12 @@ public class SelectRuneCollectionView: UICollectionView, UICollectionViewDataSou
...
     
     public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
-        return MemoryStorage.GenerationRunes.count
+        let runes = CoreDataManager.shared.fetchAllGeneratorNodes()
+        return runes.count
     }
                 
     public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
-        return self.runes[indexPath.row]
+            return self.runes[indexPath.row]
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Controllers/Generator flow/SelectRune/SelectRuneCollectionView.swift:66]: ???

File: Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift [hunk 5]
```diff
@@ -215,9 +235,16 @@ public class SelectionRuneVC: UIViewController, UIGestureRecognizerDelegate {
 
         selectRunesView.selectRune(rune: rune)
         
-        for cell in (self.selectedRunesView.visibleCells as! [SelectedRuneCell]).sorted(by: {c1, c2 in return c1.indexPath.row < c2.indexPath.row} ) {
+        for cell in (self.selectedRunesView.visibleCells as! [SelectedRuneCell])
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift:238]: Force unwrap

File: Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift [hunk 6]
```diff
@@ -215,9 +235,16 @@ public class SelectionRuneVC: UIViewController, UIGestureRecognizerDelegate {
...
+            .sorted(by: {c1, c2 in return c1.indexPath.row < c2.indexPath.row} ) {
+
             if !cell.isSelected {
-                cell.selectRune(SelectedRuneModel(title: rune.model!.title, image: rune.model!.image.image, index: rune.indexPath, id: rune.model!.id))
+                guard let imageData = rune.model?.runeImage?.image else { return }
+                guard let image = UIImage(data: imageData) else { return }
+                cell.selectRune(SelectedRuneModel(title: rune.model!.title,
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift:244]: Force unwrap

File: Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift [hunk 7]
```diff
@@ -245,9 +281,16 @@ public class SelectionRuneVC: UIViewController, UIGestureRecognizerDelegate {
             
             self.selectRunesView.selectRune(rune: rune)
             
-            for cell in (self.selectedRunesView.visibleCells as! [SelectedRuneCell]).sorted(by: {c1, c2 in return c1.indexPath.row < c2.indexPath.row} ) {
+            for cell in (self.selectedRunesView.visibleCells as! [SelectedRuneCell])
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift:284]: Force unwrap

File: Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift [hunk 8]
```diff
@@ -245,9 +281,16 @@ public class SelectionRuneVC: UIViewController, UIGestureRecognizerDelegate {
...
+                .sorted(by: {c1, c2 in return c1.indexPath.row < c2.indexPath.row} ) {
+
                 if !cell.isSelected {
-                    cell.selectRune(SelectedRuneModel(title: rune.model!.title, image: rune.model!.image.image, index: rune.indexPath, id: rune.model!.id))
+                    guard let imageData = rune.model?.runeImage?.image else { return }
+                    guard let image = UIImage(data: imageData) else { return }
+                    cell.selectRune(SelectedRuneModel(title: rune.model!.title,
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift:290]: force unwrapS

File: Runar/Runar/Controllers/Library flow/LibraryCells/LibraryRuneCell.swift [hunk 9]
```diff
@@ -10,12 +10,12 @@ import UIKit
...
     
     // MARK: - Funcs
-    public override func bind(node: LibraryNode) -> Void {
+    public override func bind(node: LibraryCoreData) -> Void {
         
-        let runeTitle: UILabel = bindRuneTitle(title: node.title!)
-        let runeImage: UIImageView = bindRuneImage(url: node.imageUrl!)
-        let tagsCV: UICollectionView = bindTagsCV(with: node.tags!)
-        let runeDesc: UILabel = bindRunDescription(description: node.content!)
+        let runeTitle: UILabel = bindRuneTitle(title: node.title ?? "")
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Controllers/Library flow/LibraryCells/LibraryRuneCell.swift:15]: Why do you use this record?

File: Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+import UIKit
+
+final class CoreDataManager {
+
+    // MARK: - Properties
+    static let shared = CoreDataManager()
+    private let nameCoreData = "CoreData"
+    var generatorIsLoaded: Bool = false
+    var libraryIsLoaded: Bool = false
+    private var context = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift:19]: Why do you use private queue?

Comment (AliaksandrNiehaj) [reply Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift:19]: And why for did you create context?

File: Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    private lazy var persistentContainer: NSPersistentContainer = {
+        let container = NSPersistentContainer(name: nameCoreData)
+        container.loadPersistentStores { _, error in
+            if let error = error {
+                debugPrint("Unresolved error \(error)")
+            }
+        }
+        return container
+    }()
+    private var parentContext: NSManagedObjectContext {
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift:29]: Why for?

File: Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+        persistentContainer.viewContext
+    }
+
+    // MARK: - Override funcs
+    private init() {
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift:34]: Why for?

**Review by AliaksandrNiehaj — CHANGES_REQUESTED** [review]

File: Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+        let allNodes = fetchAllLibraryNodes()
+
+        // Running on LibraryData array
+        // Take elements only with non-empty child arrays (childIds)
+        for structObject in libraryData where !structObject.childIds.isEmpty {
+
+            // Running through childIds of String type
+            // Find the child element in the array of core data objects (allNodes)
+            // with the corresponding id (structObjects.id)
+            for strChild in structObject.childIds {
```
Comment (AliaksandrNiehaj) [inline Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift:114]: You can use forEach()

**Review by AliaksandrNiehaj — CHANGES_REQUESTED** [review]

File: Runar/Runar/Application/AppDelegate.swift (hunk 1 above)
Comment (Qeezii) [reply Runar/Runar/Application/AppDelegate.swift]: Fixed

File: Runar/Runar/Controllers/Generator flow/GenerationPopUpViewController.swift (hunk 2 above)
Comment (Qeezii) [reply Runar/Runar/Controllers/Generator flow/GenerationPopUpViewController.swift:109]: First get imageData with Data type, then use it to get image

File: Runar/Runar/Controllers/Generator flow/SelectRune/SelectRuneCell.swift (hunk 3 above)
Comment (Qeezii) [reply Runar/Runar/Controllers/Generator flow/SelectRune/SelectRuneCell.swift:46]: GeneratorCoreData is a model for runes

File: Runar/Runar/Controllers/Generator flow/SelectRune/SelectRuneCollectionView.swift (hunk 4 above)
Comment (Qeezii) [reply Runar/Runar/Controllers/Generator flow/SelectRune/SelectRuneCollectionView.swift:66]: Re-indent command was applied

File: Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift (hunk 5 above)
Comment (Qeezii) [reply Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift]: Fixed

File: Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift (hunk 6 above)
Comment (Qeezii) [reply Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift]: Fixed

File: Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift (hunk 7 above)
Comment (Qeezii) [reply Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift]: Fixed

File: Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift (hunk 8 above)
Comment (Qeezii) [reply Runar/Runar/Controllers/Generator flow/SelectionRuneVC.swift]: Fixed

File: Runar/Runar/Controllers/Library flow/LibraryCells/LibraryRuneCell.swift (hunk 9 above)
Comment (Qeezii) [reply Runar/Runar/Controllers/Library flow/LibraryCells/LibraryRuneCell.swift:15]: Nil check and unwrap optional

File: Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift (hunk 10 above)
Comment (Qeezii) [reply Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift:19]: To work with Core Data on background thread, child context is required

File: Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift (hunk 11 above)
Comment (Qeezii) [reply Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift:29]: Parent context is required to access the Core Data persistent store

File: Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift (hunk 12 above)
Comment (Qeezii) [reply Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift:34]: To link context and parentContext

File: Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift (hunk 13 above)
Comment (Qeezii) [reply Runar/Runar/Services/Memory/CoreData/CoreDataManager.swift]: Fixed, instead "for in" used "forEach"

File: Runar/Runar/Controllers/Library flow/LibraryCells/LibraryRuneCell.swift (hunk 9 above)
Comment (Qeezii) [reply Runar/Runar/Controllers/Library flow/LibraryCells/LibraryRuneCell.swift]: Used "guard" instead "??"

**Review by AliaksandrNiehaj — APPROVED** [review]

**Review by GeraStupakov — APPROVED** [review]

**Review by dmowski — CHANGES_REQUESTED** [review]
Reduce count of file changes

**Pull request closed without merging**